### PR TITLE
If we can't squash for any reason, then simply do not optimize the items loop.

### LINKED ISSUE #15649: yum module fails with 'dict object' has no attribute '\\x00$' when used in loop and with dict lookup
##### ISSUE TYPE
- Bug Report
##### ANSIBLE VERSION

```
ansible 2.2.0 (devel f2980fc565) last updated 2016/04/28 14:44:57 (GMT +000)
  lib/ansible/modules/core: (detached HEAD e78ee3b128) last updated 2016/04/28 14:46:09 (GMT +000)
  lib/ansible/modules/extras: (detached HEAD ca310f3d15) last updated 2016/04/28 14:46:13 (GMT +000)
  config file = 
  configured module search path = Default w/o overrides
```
##### CONFIGURATION

No configuration.
##### OS / ENVIRONMENT

Reproduced on Fedora 23. I'm connecting to localhost, however the error happens before connection attempt is made.
##### SUMMARY

Running the attached playbook produces the rather cryptic error `'dict object' has no attribute '\\x00$'`. The bug appears to be related to:
- yum module (dnf, apt too),
- loops,
- dictionary lookup in the name argument to the module,

as taking any out of the equation does not reproduce the error.
##### STEPS TO REPRODUCE

Run following playbook:

```

---
- hosts: all
  gather_facts: false
  vars:
    package_map:
      foo: bar
    packages:
    - foo
  tasks:
  - yum: name={{ package_map[item] }}
    with_items: "{{ packages }}"
```

`ansible-playbook -i hosts reproduce.yml -vvvv`
##### EXPECTED RESULTS

Package `bar` to be installed using `yum`.
##### ACTUAL RESULTS

```
No config file found; using defaults
Loaded callback default of type stdout, v2.0

PLAYBOOK: reproduce.yml ********************************************************
1 plays in reproduce.yml

PLAY [all] *********************************************************************

TASK [yum] *********************************************************************
task path: /home/fedora/ansible-reproducer/reproduce.yml:12
fatal: [localhost]: FAILED! => {"failed": true, "msg": "'dict object' has no attribute '\\x00$'"}

NO MORE HOSTS LEFT *************************************************************
        to retry, use: --limit @reproduce.retry

PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   
```

### PR DESCRIPTION
**Pull request by abadger**
##### ISSUE TYPE
- Bugfix Pull Request
##### ANSIBLE VERSION

2.x
##### SUMMARY

Catch any traceback in the squashing code.

Also add more squashing testcases

Fixes #15649

### CODE REVIEW

**Comment by willthames** [comment]
Can we backport this to 2.0 also?

File: test/units/executor/test_task_executor.py [hunk 1]
```diff
@@ -230,9 +232,31 @@ def _evaluate_conditional(templar, variables):
...
+        self.assertEqual(new_items, ['a', 'b', 'c'])
+
+        # Maybe should raise an error in this case.  The user would have to specify:
+        # - yum: name="{{ packages[item] }}"
+        #   with_items:
+        #     - ['a', 'b']
+        #     - ['foo', 'bar']
+        # you can't use a list as a dict key so that would probably throw
+        # an error later.  If so, we can throw it now instead.
+        # Squashing in this case would not be intuitive as the user is being
```
Comment (willthames) [inline test/units/executor/test_task_executor.py:247]: Squashing is fine in this case, I think, if a list of lists is effectively just another way of writing a list

e.g. sometimes you might want to do:

```
yum: 
  name: "{{item}}"
  state: present
with_items:
- "{{list_of_default_rpms}}"
- "{{list_of_custom_rpms}}"
```

You can use union to join the lists of course, but I'm not sure what other use case drives this test that isn't amenable to squashing

**Comment by abadger** [comment]
I'll try cherry-picking after merging.  I'm not sure that we'll release another 2.0.x, though.

**Comment by abadger** [comment]
jimi-c found one issue at the meeting today (removed the line that called _squash_items by mistake).  I fixed.  PR is passing travis tests on my private repo.  Going to merge and cherry-pick so it's ready for tomorrow's rc2.

**Comment by abadger** [comment]
Okay, had to resolve some conflicts but this is now merged to stable-2.1 and stable-2.0 as well.

**Pull request merged**